File: src/main/java/ru/pearx/pmdumper/dumpers/DumperAdvancements.java
package ru.pearx.pmdumper.dumpers;

import net.minecraft.advancements.Advancement;
import net.minecraft.advancements.AdvancementList;
import net.minecraft.advancements.AdvancementManager;
import net.minecraft.advancements.DisplayInfo;
import net.minecraft.world.World;
import net.minecraft.world.WorldServer;
import net.minecraftforge.common.DimensionManager;
import net.minecraftforge.fml.common.registry.ForgeRegistries;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import ru.pearx.pmdumper.dumpers.base.IDumper;
import ru.pearx.pmdumper.dumpers.base.PMDData;

import java.util.*;

/**
 * Created by mrAppleXZ on 14.06.17 12:11.
 */
@SideOnly(Side.CLIENT)
public class DumperAdvancements implements IDumper
{

    @Override
    public String getName()
    {
        return "advancements";
    }

    @Override
    public PMDData getData()
    {
        List<Advancement> advs = new ArrayList<>();
        for(WorldServer wrld : DimensionManager.getWorlds())
        {
            for(Advancement adv : wrld.getAdvancementManager().getAdvancements())
            {
                if(!advs.contains(adv))
                    advs.add(adv);
            }
        }

        List<List<String>> data = new ArrayList<>();
        Map<String, Integer> counts = new HashMap<>();
        data.add(Arrays.asList("ID", "Display Text", "Display Info", "Parent", "Children", "Rewards"));
        for(Advancement adv : advs)
        {
            StringBuilder children = new StringBuilder();
            for(Advancement ch : adv.getChildren())
            {
                children.append(ch.getId());
                children.append(System.lineSeparator());
            }
            if(children.length() >= System.lineSeparator().length())
                children.delete(children.length() - System.lineSeparator().length(), children.length());
            data.add(Arrays.asList(adv.getId().toString(),
                    adv.getDisplayText().getFormattedText(),
                    adv.getDisplay() == null ? "no" : displayInfoToString(adv.getDisplay()),
                    adv.getParent() == null ? "no" : adv.getParent().getId().toString(),
                    children.toString(),
                    adv.getRewards().toString()
            ));
            PMDData.plusCounts(counts, adv.getId().getResourceDomain(), 1);
        }
        return new PMDData(data, counts);
    }

    public String displayInfoToString(DisplayInfo inf)
    {
        StringBuilder bld = new StringBuilder();
        bld.append("{");
        bld.append("title=").append(inf.getTitle().getFormattedText()).append(", ").append(System.lineSeparator());
        bld.append("description=").append(inf.getDescription().getFormattedText()).append(", ").append(System.lineSeparator());
        bld.append("icon=").append(inf.getIcon().toString()).append(", ").append(System.lineSeparator());
        bld.append("x=").append(inf.getX()).append(", ").append(System.lineSeparator());
        bld.append("y=").append(inf.getX()).append(", ").append(System.lineSeparator());
        bld.append("background=").append(inf.getBackground() == null ? "no" : inf.getBackground().toString()).append(", ").append(System.lineSeparator());
        bld.append("frame=").append(inf.getFrame().toString()).append(", ").append(System.lineSeparator());
        bld.append("hidden=").append(inf.isHidden()).append(", ").append(System.lineSeparator());
        bld.append("announce=").append(inf.shouldAnnounceToChat()).append(", ").append(System.lineSeparator());
        bld.append("toast=").append(inf.shouldShowToast());
        bld.append("}");
        return bld.toString();
    }
}

File: src/main/kotlin/ru/pearx/pmdumper/dumper/standard/DumperAdvancements.kt
@file:JvmMultifileClass
@file:JvmName("StandardDumpers")
package ru.pearx.pmdumper.dumper.standard

import net.minecraft.advancements.Advancement
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.DimensionManager
import ru.pearx.pmdumper.*
import ru.pearx.pmdumper.dumper.dumper
import ru.pearx.pmdumper.utils.*

val DumperAdvancements = dumper {
    registryName = ResourceLocation(ID, "advancements")
    header = listOfNotNull("ID", "Display Text", "Title", "Description", client("Icon"), client("X"), client("Y"), client("Background"), "Frame", "Is Hidden", "Should Announce", client("Should Show Toast"), "Parent", "Children", "Reward Experience", "Reward Loot", "Reward Recipes", "Reward Function")
    amounts {
        for (adv in collectAdvancements())
            this += adv.id
    }
    iterator {
        for (adv in collectAdvancements()) {
            with(ArrayList<String>(header.size)) {
                with(adv) {
                    add(id.toString())
                    add(displayText.formattedText)
                    display?.run {
                        add(title.formattedText)
                        add(description.formattedText)
                        client {
                            add(icon.toFullString())
                            add(x.toString())
                            add(y.toString())
                            add(background?.toPath() ?: "")
                        }
                        add(frame.toString())
                        add(isHidden.toPlusMinusString())
                        add(shouldAnnounceToChat().toPlusMinusString())
                        client { add(shouldShowToast().toPlusMinusString()) }
                    } ?: repeat(if (isClient) 10 else 5) { add("") }
                    add(parent?.id?.toString() ?: "")
                    add(StringBuilder().apply {
                        var start = true
                        for (child in children) {
                            if (start)
                                start = false
                            else
                                appendln()
                            append(child.id.toString())
                        }
                    }.toString())
                    rewards.apply {
                        add(experience.toString())
                        add(loot.joinToString(separator = System.lineSeparator()))
                        add(recipes.joinToString(separator = System.lineSeparator()))
                        add(function?.toString() ?: "")
                    }
                }
                yield(this)
            }
        }
    }
}

private fun collectAdvancements(): List<Advancement> {
    val advancements = mutableListOf<Advancement>()
    for (w in DimensionManager.getWorlds())
        for (adv in w.advancementManager.advancements)
            if (adv !in advancements)
                advancements.add(adv)
    return advancements
}